### ISSUE TITLE
Failure to restart after update to new Eclipse Version

### ISSUE THREAD

**Issue by wimjongman** [eclipse-equinox/equinox#22]

**Comment by laeubi** [comment]
> > > Yes, but better slow start than crashing app :-\
> > 
> > 
> > Can we add -clean only after an update?
> 
> That's my earlier point. I thought equinox was already behaving like a -clean when the set of bundles changes...

One should keep in mind that `-clean` is not only "making restart slow" it actually wipe out the OSGi Storage area that for example contains the bundle data storage! So this would be fatal to having equinox doing it "by default", because even Eclipse(!) stores most data in the Workspace this is not true for standard OSGi bundles!

**Comment by tjwatson** [comment]
> > > Yes, but better slow start than crashing app :-\
> > 
> > 
> > Can we add -clean only after an update?
> 
> That's my earlier point. I thought equinox was already behaving like a -clean when the set of bundles changes...

Equinox does not do this.  I'm not sure if it is a good idea or not for p2 to force a -clean on first restart after update.  Some solution would need to be coded that makes that happen only on first restart after update.  I imagine it will disturb some to have all their bundles data area wiped clean even if their bundles are not the ones getting updates.

**Comment by wimjongman** [comment]
Yes, can do. Please hang on.

**Comment by wimjongman** [comment]
To reproduce, download an Eclipse RCP from here: https://miworkplace.com/index.php?content=download
then unzip, run, and check for updates.

**Comment by laeubi** [comment]
@tjwatson I also see this with "regular" bundles, more bad, if it contains native code this sometimes even crash the whole process..

> What p2 should be doing is asking the framework to refresh all the bundles it uninstalled forcing the fragments to re-resolve against the newly installed bundle

This might be if you do "Apply now" but I think the case I and wim encounter is more the "restart now", then the application returns a special return code (APP_RESTART) that is then cached and the app is started immediately again.

**Comment by tjwatson** [comment]
"restart now" is the default and what most will use.  That should be using this change:

https://github.com/eclipse-equinox/p2/commit/907ec00a829e2fe03f156089775c70ddac834d6e

The intention there was to force ALL bundles to be refreshed when we detect that simpleconfigurator is doing the update operation during a restart.  Something must be going wrong with that fix that I put in for https://bugs.eclipse.org/bugs/show_bug.cgi?id=565066

It would be nice to have steps I can use to reproduce ... just saw Wim post something for that.  Thanks.

**Comment by merks** [comment]
I see about the about before restarting:

![image](https://user-images.githubusercontent.com/208716/174088384-7a9e6396-2e70-45ad-85f1-ed9dbb8346a9.png)

I see these in the plugins folder after manually exiting (not restart); the fragments are newly installed.

![image](https://user-images.githubusercontent.com/208716/174088536-db38368e-8ced-4097-861a-b1cfad0b42b6.png)

I see these in the profile:
```
    <unit id='org.eclipse.equinox.common' version='3.16.100.v20220315-2327' generation='2'>
    <unit id='org.eclipse.equinox.common.nl_de' version='4.22.0.v20211218020001'>
    <unit id='org.eclipse.equinox.common.nl_es' version='4.22.0.v20211218020001'>
    <unit id='org.eclipse.equinox.common.nl_fr' version='4.22.0.v20211218020001'>
    <unit id='toolingwin32.win32.x86_64org.eclipse.equinox.common' version='1.29.0.202205171359' singleton='false'>
```
When I start from this state it fails as Win shows.  When I start again a second time, then it works.

So it seems everything is correctly installed and at least the profile reflects the correct state.  It really does look like some type of "update the modules" problem...

**Comment by laeubi** [comment]
By the way many thanks @wimjongman for reproducer and @tjwatson and @merks for looking into this, because this has bugged me in the past ranging from requiring a a manual restart up to constant crashing eclipse and only `-clean` helped to get out of the problem.

**Comment by merks** [comment]
It's rather odd though because I constantly update IDEs and Jonah tests that the previous EPP releases can be updated to the latest release/milestone so this is a pretty well tested path.  Likely something funky with the fragments is involved here...

**Comment by laeubi** [comment]
Well as you said the IDE is constantly updated so issues for *that* setup are quite easy spotted, but with a custom (RCP?) application there is not just the IDE but also other bundles that e.g. start *before* the simple configuration, having custom start levels, or even autostarts configured as well as some might even be not happy about beeing refreshed just after restart and so on...

**Comment by tjwatson** [comment]
The issue in this case is that simpleconfigurator has fragments and simpleconfigurator is asking that its own resolved fragments get refreshed because simpleconfigurator uninstalled its own fragments.  Then simpleconfigurator asks the framework to refresh and simpleconfigurator blocks (while it is starting) waiting for the async refresh packages to complete.  This causes issues because that async thread now wants to re-resolve simpleconfigurator in order to detach the fragments that got uninstalled.

To re-resolve simpleconfigurator the framework needs to grab simpleconfigurator's state lock, but that lock is being held by thead starting simpleconfigurator (which is blocked waiting for the async refresh operation).  Ultimately this causes the refresh operation to stop short before it has had a chance to do what it needs to refresh.

I think org.eclipse.equinox.internal.simpleconfigurator.ConfigApplier.getDoNotRefresh() needs fixed to not pass any fragments of simpleconfigurator similar to what it does for the systembundle.

**Comment by wimjongman** [comment]
I downloaded the SDK 4.23 and installed the Babel fragments for Eclipse from: https://download.eclipse.org/technology/babel/update-site/latest/

After installation restart was good.

Then I changed the download locations to ../latest and "Check for updates". 

This also failed to restart. So the guess that it has something to do with fragments seems more and more likely.


![image](https://user-images.githubusercontent.com/180969/174106514-2ae521f1-f2a1-4dd5-8279-5f196748791b.png)

[1655393393512.log](https://github.com/eclipse-equinox/equinox/files/8920302/1655393393512.log)

**Comment by tjwatson** [comment]
> Well as you said the IDE is constantly updated so issues for _that_ setup are quite easy spotted, but with a custom (RCP?) application there is not just the IDE but also other bundles that e.g. start _before_ the simple configuration, having custom start levels, or even autostarts configured as well as some might even be not happy about beeing refreshed just after restart and so on...

This is likely to be some bad news to hear.  But any configuration that uses simpleconfigurator and that said configuration tries to have bundles start BEFORE simpleconfigurator is not going to be happy.  simpleconfigurator assumes it is THE first bundle started and is doing its operations while nothing is running.  Everything else should have a start-level greater than simpleconfigurator if you want to have any hope of control and understanding when p2 is doing an update.

**Comment by laeubi** [comment]
> This is likely to be some bad news to hear. But any configuration that uses simpleconfigurator and that said configuration tries to have bundles start BEFORE simpleconfigurator is not going to be happy. simpleconfigurator assumes it is THE first bundle started

Actually all bundles you configure before or at the same start level are added to the config.ini and simpleconfigurator should most probably know that, I'm a bit curious that so many "hacks" are required to support simpleconfigurator, why does the refresh not happen when the application is already down?

> want to have any hope of control and understanding when p2 is doing an update

That's why in the past I only used P2 to actually collect the bundles, switch to startlevel-1 and update all bundles including refresh and return to the application start level, but actually I though I should not do too much myself and let P2 do the job ...
